tokens/hmac.cpp

```diff
@@ -87,34 +87,36 @@ void innerhash(HMACKey_d key, State_d state, Integer innerhashresult[8]) {
 
   // a single 1 bit, followed by 0's
   // The state is 1536 bits long.  Key block is 512.  Total is 2048 bits
-  message[4][0] = Integer(32, -2147483648, PUBLIC); //0x80000000;
-  message[4][1] = Integer(32, 0, PUBLIC); //0x00000000;
-  message[4][2] = Integer(32, 0, PUBLIC); //0x00000000;
-  message[4][3] = Integer(32, 0, PUBLIC); //0x00000000;
-  message[4][4] = Integer(32, 0, PUBLIC); //0x00000000;
-  message[4][5] = Integer(32, 0, PUBLIC); //0x00000000;
-  message[4][6] = Integer(32, 0, PUBLIC); //0x00000000;
-  message[4][7] = Integer(32, 0, PUBLIC); //0x00000000;
-  message[4][8] = Integer(32, 0, PUBLIC); //0x00000000;
-  message[4][9] = Integer(32, 0, PUBLIC); //0x00000000;
-  message[4][10] = Integer(32, 0, PUBLIC); //0x00000000;
-  message[4][11] = Integer(32, 0, PUBLIC); //0x00000000;
-  message[4][12] = Integer(32, 0, PUBLIC); //0x00000000;
-  message[4][13] = Integer(32, 0, PUBLIC); //0x00000000;
-  message[4][14] = Integer(32, 0, PUBLIC); //0x00000000;
-  message[4][15] = Integer(32, 2048, PUBLIC); //0x000003a0;
-
-  computeSHA256_5d(message, innerhashresult);
+//  message[4][0] = Integer(32, -2147483648, PUBLIC); //0x80000000;
```

Comment:
I'm a little surprised this doesn't cause some kind of memory issue. If these are all the same zero, why don't they write over each other when it runs? I won't question it if your tests work, but it's not the behavior I anticipated.

Reply:
Using the same input just causes the places where it's used to use the same input wires, but an input wire can be connected to multiple gates without any problem. It also never overrides a wire, it just inputs its value to the gate and puts the output on another wire. So, if the same input wire is used multiple times, it will always take the same value as input, but will never override it. Not sure how it is done in the semihonest case, but I suppose similarly. Isn't this also expected behavior because this is also possible when doing normal programming? Or am I misunderstanding your question?

Reply:
I guess it does make more sense to me in the ag2pc case, where we actually build a circuit. I'm thinking about this more in the sh2pc case, where I think it doesn't build a circuit, just executes the code on-the-fly. In that case, you might have a situation where two different variables point to the same piece of memory, so changing one variable affects the other.

I think you can avoid it by forcing copy-by-value rather than -by-reference. Or maybe these aren't pointers and do a deep copy by default.

As long as the sh2pc tests still run, I won't complain though.

Reply:
Well, the tests still run. And I'm not a C++ developer, but I would assume this would be a copy by value, no? Isn't that always the case in an assignment, unless I specifically assign the reference to the object to a new variable?